walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,123 @@
 
 package com.google.sps;
 
+import java.util.Arrays;
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public static final int DAY_START = TimeRange.START_OF_DAY;
+    public static final int DAY_END = TimeRange.END_OF_DAY;
+    public static final TimeRange WHOLE_DAY = TimeRange.WHOLE_DAY;
+
+    /**
+      * Return a collection of TimeRanges that attendees can meet for a requested 
+      * meeting. If there are no TimeRanges that works for both required and optional 
+      * attendees, find TimeRanges that just work for required attendees. If there 
+      * are no required attendees, find the TimeRanges for just the optional attendees.
+      */
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        long meetingDuration = request.getDuration();
+        Collection<String> requiredAttendees = request.getAttendees(); 
+        Collection<String> allAttendees = new ArrayList<>(requiredAttendees);
+        allAttendees.addAll(request.getOptionalAttendees());
+
+        if (meetingDuration > WHOLE_DAY.duration() || meetingDuration <= 0) {
+            // The meeting duation is invalid.
+            return Arrays.asList();
+        }
+
+        // Find TimeRanges when both required and optional attendees are available.
+        Collection<TimeRange> allMeetingTimes = 
+            findMeetingTimes(events, allAttendees, meetingDuration);
+
+        if (!allMeetingTimes.isEmpty()) {
+            return allMeetingTimes;
+        }
+
+        // If there is no required attendee, return an empty collection.
+        if (requiredAttendees.isEmpty()) {
+            return Arrays.asList();
+        }
```

Comment:
```suggestion
        if (requiredAttendees.isEmpty()) {
            // There were no required attendees (all were optional) and we failed
            // to find an available meeting time so we return an empty result
            // because there is nothing else we can do.
            return Arrays.asList();
        }
```

Reply:
The comment is updated! :)